File: Sorting/Sorting/SortBase.cs
﻿namespace Sorting
{
    using System;
    using System.Linq;
    using System.Collections.Generic;
    using System.Diagnostics;

	public abstract class SortBase
	{
		private Stopwatch _stopWatch;
		private readonly RankingSystem _rankingSystem;
		private readonly string _sortName;

		protected SortBase(RankingSystem rankingSystem)
		{
			_rankingSystem = rankingSystem;
			_sortName = GetType().Name;
		}
		
		public List<ICollection<int>> IntermediateSorts { get; } = new List<ICollection<int>>();

		public int? Rank => _rankingSystem.Get(_sortName);

		public long Duration { get; private set; }

		public virtual IEnumerable<int> Execute(ICollection<int> unsortedList)
		{
			_stopWatch = Stopwatch.StartNew();
			IntermediateSorts.Clear();
			return unsortedList;
		}

		protected void StopWatch()
		{
			_stopWatch.Stop();
			Duration = _stopWatch.ElapsedTicks;
			_rankingSystem.Add(_sortName, Duration);
		}

		protected IEnumerable<int> Copy(IEnumerable<int> input)
		{
			foreach (var item in input)
			{
				yield return item;
			}
		}

        protected void AddToIntermediateSorts(IEnumerable<int> intermediatSort)
        {
            IntermediateSorts.Add(Copy(intermediatSort).ToList());
        }

        protected void Log(ICollection<int> collection)
        {
            Console.WriteLine("[" + string.Join(", ", collection) + "]");
        }
	}
}

File: Sorting/Sorting.WebPure/src/scripts/Sorting/Algorithms/SortBase.ts
namespace Sorting.Algorithms {
	import RankingSystem = Tools.RankingSystem;

	export class SortBase {
		private rankingSystem: RankingSystem;
		private intermediateSorts: number[][];
		private start: number;
		private end: number;
		private name: string;

		constructor(rankingSystem: RankingSystem) {
			this.rankingSystem = rankingSystem;
			this.name = this.constructor.name;
		}

		public getRank(): number {
			return this.rankingSystem.getRank(this.name);
		}

		protected copy(array: number[]): number[] {
			const newArray: number[] = [];
			for (let i = 0; i < array.length; i++) {
				newArray[i] = array[i];
			}

			return newArray;
		}

		protected addToIntermediateSorts(intermediateSort: number[]) {
			this.intermediateSorts.push(this.copy(intermediateSort));
		}

		public execute(unsortedList: number[]): number[] {
			this.start = new Date().getTime();
			this.intermediateSorts = [];
			return unsortedList;
		}

		public getIntermediateSorts(): number[][] {
			return this.intermediateSorts;
		}

		protected stopWatch(): void {
			this.end = new Date().getTime();
			this.rankingSystem.add(this.name, this.getDuration());
		}

		public getDuration(): number {
			return this.end - this.start;
		}
	}
}